core/src/main/scala-2/chisel3/Contract.scala

```diff
@@ -0,0 +1,114 @@
+    * val b = Contract((a<<3)+a) { b =>
+    *   EnsureProperty(b === a*9)
+    * }
+    *
+    * // Contract with multiple arguments:
+    * //   "carry-save adder reduces number of terms but doesn't change sum"
+    * // During formal verification, this is interpreted as:
+    * // - an assertion that the carry-save adder indeed compresses the three
+    * //   terms p/q/r down to only two terms u/v which have the same sum
+    * // - symbolic values u/v constrained to sum up to the same value as p/q/r,
+    * //   as a simplification of other proofs
+    * val csa_s = p^q^r
+    * val csa_c = (p&q | (p^q)&r) << 1
+    * val (u,v) = Contract(csa_c, csa_s) { case (u,v) =>
+    *   EnsureProperty(u+v === p+q+r)
+    * }
+    * }}}
+    */
+  def apply(head: Data, tail: Data*): (Any => Unit) => Any = macro apply_impl
+
+  // This uses a macro internally to allow for contracts with a single argument
+  // and contracts with a tuple of arguments to be created. The macro packages
+  // the sequence of arguments up into a tuple (`Tuple1` for single arguments,
+  // and the corresponding `TupleN` for multiple arguments). Without this macro,
+  // the contracts would always operate on tuples (`Product`), forcing the user
+  // to use `val Tuple1(a) = Contract(Tuple1(b)) { case Tuple1(x) => ... }` for
+  // contracts with only a single argument.
+  def apply_impl(c: whitebox.Context)(head: c.Expr[Data], tail: c.Expr[Data]*): c.Expr[(Any => Unit) => Any] = {
+    import c.universe._
+    val args = head +: tail
+    val mapping =
+      q"(values => (..${args.zipWithIndex.map { case (arg, i) => q"values(${i}).asInstanceOf[${arg.tree.tpe}]" }}))"
+    val result = q"Contract.mapped(Seq(..${args.map(_.tree)}), $mapping)(_)"
+    c.Expr[(Any => Unit) => Any](result)
+  }
+
+  /** Create a contract from a sequence of expressions.
+    *
+    * The `mapping` function is used to convert the `Seq[Data]` to a
+    * user-defined type `R`, which is then passed to the contract body and
+    * returned as a result.
+    */
+  def mapped[R](args: Seq[Data], mapping: Seq[Data] => R)(body: R => Unit)(implicit sourceInfo: SourceInfo): R = {
+    // Create a sequence of contract results, one for each argument.
+    val results = args.map { arg =>
+      val result = arg.cloneTypeFull
+      experimental.requireIsChiselType(result, "contract result type")
+      result.bind(binding.OpBinding(Builder.forcedUserModule, Builder.currentBlock))
+      result
+    }
+
+    // Create the contract.
+    val contract = Builder.pushCommand(new DefContract(sourceInfo, results, args.map(_.ref(sourceInfo))))
+
+    // Map the sequence of results to the user-defined type `R`, pass it to the
+    // body to build the contract body, and also return it as the result of the
+    // contract as a whole.
+    val mapped = mapping(results)
+    Builder.forcedUserModule.withRegion(contract.region)(layer.elideBlocks(body(mapped)))
+    mapped
+  }
+}
```

Comment:
What about Scala 3? The code that can be shared should be in `src/main/scala`. The macro is cool and I do think it's better, but if you just did the annoying thing and write the method for 1 to N arguments (e.g. [DataProduct support for Tuples](https://github.com/chipsalliance/chisel/blob/e20069ee8fd73ffb036a2a2cc17c7b3a9956dd3c/core/src/main/scala/chisel3/experimental/dataview/DataProduct.scala#L125)), then it would work for both Scala 2 and 3.

Reply:
Good call. I've move the common code into `scala/chisel3` and added an explicit `scala-3/chisel3` with the Scala 3 specific stuff.